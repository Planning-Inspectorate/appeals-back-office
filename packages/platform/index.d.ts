import { HandlerFunction } from 'got';

declare module '@pins/platform' {
	export function createTtlHandler(): HandlerFunction;
	export function validatePostcode(value: string): boolean;
	export function validatePastDate(value: Date | string | number): boolean;
	export function validateFutureDate(value: Date | string | number): boolean;
	export function yesterday(): Date;
}

export interface LocalPlanningDepartment {
	FID: number;
	LPA21CD: string;
	LPA21NM: string;
	Co_terminous: 'Y' | 'N';
}
