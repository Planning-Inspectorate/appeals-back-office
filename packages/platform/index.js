export * from './http/ttl.js';
export * from './util/date.js';
export * from './validators/date.js';
export * from './validators/postcode.js';
