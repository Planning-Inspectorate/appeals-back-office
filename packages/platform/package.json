{
  "name": "@pins/platform",
  "version": "0.0.0",
  "type": "module",
  "main": "index.js",
  "private": true,
  "engines": {
    "node": ">=22.0.0 <= 22.14.0"
  },
  "scripts": {
    "clean": "rimraf .turbo node_modules dist",
    "lint:js": "npx eslint .",
    "lint:js:fix": "npx eslint . --fix",
    "test": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci",
    "test:cov": "cross-env NODE_ENV=test NODE_OPTIONS=\"--experimental-vm-modules --expose-gc\" npx jest --logHeapUsage --ci --coverage",
    "tscheck": "npx tsc -p jsconfig.json --maxNodeModuleJsDepth 0"
  },
  "dependencies": {
    "date-fns": "2.28.0",
    "dotenv": "16.0.3",
    "got": "^12.5.2",
    "html": "^1.0.0",
    "lodash-es": "^4.17.21",
    "node-html-parser": "^5.3.3",
    "randexp": "^0.5.3"
  },
  "devDependencies": {
    "@types/html": "^1.0.4",
    "@types/lodash-es": "^4.17.12",
    "cross-env": "^7.0.3",
    "jest": "^29.7.0",
    "type-fest": "^4.37.0"
  }
}
