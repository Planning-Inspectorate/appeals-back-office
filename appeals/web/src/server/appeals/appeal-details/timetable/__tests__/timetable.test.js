import { parseHtml } from '@pins/platform';
import nock from 'nock';
import supertest from 'supertest';
import { appealData as baseAppealData } from '#testing/app/fixtures/referencedata.js';
import { createTestEnvironment } from '#testing/index.js';

const { app, teardown } = createTestEnvironment();
const request = supertest(app);
const baseUrl = '/appeals-service/appeal-details/1/timetable';

describe('Appeal Timetables', () => {
	afterEach(teardown);

	it('should render "Timetable due dates" page for householder appeal type', async () => {
		const appealData = {
			...baseAppealData,
			appealTimetable: {
				appealTimetableId: 1
			}
		};

		nock('http://test/').get('/appeals/1').reply(200, appealData);

		const response = await request.get(`${baseUrl}/edit`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
		expect(element.innerHTML).toContain('LPA questionnaire due</legend>');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-day');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-month"');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-year"');
		expect(element.innerHTML).toContain('Continue</button>');
	});

	it('should render correct "Timetable due dates" page for S78 appeal type and status lpa_questionnaire', async () => {
		const appealData = {
			...baseAppealData,
			appealTimetable: {
				appealTimetableId: 1
			}
		};
		appealData.appealType = 'Planning appeal';
		appealData.appealStatus = 'lpa_questionnaire';

		nock('http://test/').get('/appeals/1').reply(200, appealData);

		const response = await request.get(`${baseUrl}/edit`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
		expect(element.innerHTML).toContain('Timetable due dates</h1>');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-day');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-month"');
		expect(element.innerHTML).toContain('name="lpa-questionnaire-due-date-year"');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-day');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-month"');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-year"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-day');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-year"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-day');
		expect(element.innerHTML).toContain('name="final-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-year"');
		expect(element.innerHTML).toContain('Continue</button>');
	});

	it('should render correct "Timetable due dates" page for S78 appeal type and status lpa_questionnaire and lpaq received', async () => {
		const appealData = {
			...baseAppealData,
			appealTimetable: {
				appealTimetableId: 1
			}
		};
		appealData.appealType = 'Planning appeal';
		appealData.appealStatus = 'lpa_questionnaire';
		appealData.documentationSummary = {
			lpaQuestionnaire: {
				status: 'received',
				dueDate: '2024-10-11T10:27:06.626Z',
				receivedAt: '2024-08-02T10:27:06.626Z',
				representationStatus: 'awaiting_review'
			}
		};

		nock('http://test/').get('/appeals/1').reply(200, appealData);

		const response = await request.get(`${baseUrl}/edit`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
		expect(element.innerHTML).toContain('Timetable due dates</h1>');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-day');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-month"');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-year"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-day');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-year"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-day');
		expect(element.innerHTML).toContain('name="final-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-year"');
		expect(element.innerHTML).toContain('Continue</button>');
	});

	it('should render correct "Timetable due dates" page for S78 appeal type and status statements', async () => {
		const appealData = {
			...baseAppealData,
			appealTimetable: {
				appealTimetableId: 1
			}
		};
		appealData.appealType = 'Planning appeal';
		appealData.appealStatus = 'statements';

		nock('http://test/').get('/appeals/1').reply(200, appealData);

		const response = await request.get(`${baseUrl}/edit`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
		expect(element.innerHTML).toContain('Timetable due dates</h1>');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-day');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-month"');
		expect(element.innerHTML).toContain('name="lpa-statement-due-date-year"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-day');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="ip-comments-due-date-year"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-day');
		expect(element.innerHTML).toContain('name="final-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-year"');
		expect(element.innerHTML).toContain('Continue</button>');
	});

	it('should render correct "Timetable due dates" page for S78 appeal type and status final_comments', async () => {
		const appealData = {
			...baseAppealData,
			appealTimetable: {
				appealTimetableId: 1
			}
		};
		appealData.appealType = 'Planning appeal';
		appealData.appealStatus = 'final_comments';

		nock('http://test/').get('/appeals/1').reply(200, appealData);

		const response = await request.get(`${baseUrl}/edit`);
		const element = parseHtml(response.text);

		expect(element.innerHTML).toMatchSnapshot();
		expect(element.innerHTML).toContain('Final comments due</legend>');
		expect(element.innerHTML).toContain('name="final-comments-due-date-day');
		expect(element.innerHTML).toContain('name="final-comments-due-date-month"');
		expect(element.innerHTML).toContain('name="final-comments-due-date-year"');
		expect(element.innerHTML).toContain('Continue</button>');
	});
});
