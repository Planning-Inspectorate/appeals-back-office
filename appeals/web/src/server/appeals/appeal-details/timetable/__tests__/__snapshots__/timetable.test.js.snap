// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Appeal Timetables should render "Timetable due dates" page for householder appeal type 1`] = `
"<main class="govuk-main-wrapper" id="main-content">
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-two-thirds"><span class="govuk-caption-l">Appeal 351062</span>
            <h1 class="govuk-heading-l">Timetable due dates</h1>
        </div>
    </div>
    <div class="govuk-grid-row">
        <div class="govuk-grid-column-full"></div>
    </div>
    <form method="post" novalidate="novalidate">
        <div class="govuk-form-group">
            <fieldset class="govuk-fieldset" role="group" aria-describedby="lpa-questionnaire-due-date-hint">
                <legend class="govuk-fieldset__legend govuk-fieldset__legend--m">LPA questionnaire due</legend>
                <div id="lpa-questionnaire-due-date-hint"
                class="govuk-hint">For example, 27 3 2007</div>
                <div class="govuk-date-input" id="lpa-questionnaire-due-date">
                    <div class="govuk-date-input__item">
                        <div class="govuk-form-group">
                            <label class="govuk-label govuk-date-input__label" for="lpa-questionnaire-due-date-day">Day</label>
                            <input class="govuk-input govuk-date-input__input govuk-input--width-2"
                            id="lpa-questionnaire-due-date-day" name="lpa-questionnaire-due-date-day"
                            type="text" inputmode="numeric">
                        </div>
                    </div>
                    <div class="govuk-date-input__item">
                        <div class="govuk-form-group">
                            <label class="govuk-label govuk-date-input__label" for="lpa-questionnaire-due-date-month">Month</label>
                            <input class="govuk-input govuk-date-input__input govuk-input--width-2"
                            id="lpa-questionnaire-due-date-month" name="lpa-questionnaire-due-date-month"
                            type="text" inputmode="numeric">
                        </div>
                    </div>
                    <div class="govuk-date-input__item">
                        <div class="govuk-form-group">
                            <label class="govuk-label govuk-date-input__label" for="lpa-questionnaire-due-date-year">Year</label>
                            <input class="govuk-input govuk-date-input__input govuk-input--width-4"
                            id="lpa-questionnaire-due-date-year" name="lpa-questionnaire-due-date-year"
                            type="text" inputmode="numeric">
                        </div>
                    </div>
                </div>
            </fieldset>
        </div>
        <div class="govuk-button-group">
            <button type="submit" class="govuk-button" data-module="govuk-button">Continue</button>
        </div>
    </form>
</main>"
`;
